### PR TITLE
[RLlib] Auto-framework, retire `use_pytorch` in favor of `framework=...`

### PR DESCRIPTION
**Pull request by sven1977**
This PR retires the `use_pytorch` config setting in favor of a more generic `framework=tf|tfe|torch|auto` setup, where `auto` causes torch to be used iff only PyTorch is installed (otherwise: tf).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/agents/dqn/simple_q_tf_policy.py [hunk 1]
```diff
@@ -55,15 +55,15 @@ def build_q_models(policy, obs_space, action_space, config):
         action_space=action_space,
         num_outputs=action_space.n,
         model_config=config["model"],
-        framework="torch" if config["use_pytorch"] else "tf",
+        framework="torch" if config["framework"] == "torch" else "tf",
```
Comment (ericl) [inline rllib/agents/dqn/simple_q_tf_policy.py:58]: Can these be `framework=config["framework"]`?

File: rllib/agents/ppo/ppo.py [hunk 2]
```diff
@@ -185,12 +183,12 @@ def validate_config(config):
         logger.warning(
             "Using the simple minibatch optimizer. This will significantly "
             "reduce performance, consider simple_optimizer=False.")
-    elif config["use_pytorch"] or (tf and tf.executing_eagerly()):
+    elif config["framework"] != "torch":
```
Comment (ericl) [inline rllib/agents/ppo/ppo.py:186]: Shouldn't this be `config["framework"] in ["torch", "tfe"]`?

File: rllib/agents/ppo/tests/test_ppo.py [hunk 3]
```diff
@@ -164,7 +164,7 @@ def get_value():
             init_std = get_value()
             assert init_std == 0.0, init_std
 
-            if fw == "tf" or fw == "eager":
+            if fw != "torch":
```
Comment (ericl) [inline rllib/agents/ppo/tests/test_ppo.py:167]: Let's try to be future proof here, `if fw in ["tf", "tfe"]:`

File: rllib/agents/sac/sac_tf_policy.py [hunk 4]
```diff
@@ -63,8 +63,9 @@ def build_sac_model(policy, obs_space, action_space, config):
         action_space=action_space,
         num_outputs=num_outputs,
         model_config=config["model"],
-        framework="torch" if config["use_pytorch"] else "tf",
-        model_interface=SACTorchModel if config["use_pytorch"] else SACTFModel,
+        framework="torch" if config["framework"] == "torch" else "tf",
```
Comment (ericl) [inline rllib/agents/sac/sac_tf_policy.py:66]: Can we pass through framework directly?

File: rllib/utils/framework.py [hunk 5]
```diff
@@ -9,6 +9,17 @@
...
 
 
+def get_auto_framework():
+    """Returns a framework (str) to use automatically.
+
+    If only PyTorch is installed, returns "torch", otherwise "tf".
+    """
+    # PyTorch is installed.
+    if torch is not None and tf is None:
+        return "torch"
```
Comment (ericl) [inline rllib/utils/framework.py:19]: This should probably raise an error instead if both are installed.

File: rllib/utils/framework.py [hunk 6]
```diff
@@ -9,6 +9,17 @@
 TensorType = Any
 
 
+def get_auto_framework():
+    """Returns a framework (str) to use automatically.
+
+    If only PyTorch is installed, returns "torch", otherwise "tf".
```
Comment (ericl) [inline rllib/utils/framework.py:15]: Can we also INFO log on auto detection? This could be very confusing otherwise.

**Review by ericl — CHANGES_REQUESTED** [review]
Feel free to not do this, but could we write the code to be robust to other framework values as well, e.g., jax?

File: rllib/agents/dqn/simple_q_tf_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/dqn/simple_q_tf_policy.py:58]: Good idea. Then we have to change the ModelCatalog to be more flexible wrt tf/tfe.

File: rllib/agents/ppo/ppo.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:186]: Good catch. I probably meant `!= "tf"`.

File: rllib/agents/ppo/tests/test_ppo.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/ppo/tests/test_ppo.py:167]: +1

File: rllib/agents/sac/sac_tf_policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/agents/sac/sac_tf_policy.py:66]: +1

File: rllib/utils/framework.py (hunk 5 above)
Comment (sven1977) [reply rllib/utils/framework.py:19]: +1

File: rllib/utils/framework.py (hunk 6 above)
Comment (sven1977) [reply rllib/utils/framework.py:15]: +1

File: rllib/agents/dqn/simple_q_tf_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/dqn/simple_q_tf_policy.py:58]: done

File: rllib/agents/ppo/ppo.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:186]: done

File: rllib/agents/ppo/tests/test_ppo.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/ppo/tests/test_ppo.py:167]: done

File: rllib/agents/sac/sac_tf_policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/agents/sac/sac_tf_policy.py:66]: done, also in all similar cases

File: rllib/utils/framework.py (hunk 6 above)
Comment (sven1977) [reply rllib/utils/framework.py:15]: done

**Review by ericl — APPROVED** [review]

**Pull request merged**